### ISSUE TITLE
MKL-DNN causes wrong results on Xeon

### ISSUE THREAD

**Issue by fracting** [pytorch/pytorch#18218]
Editorial note: We think this is an MKL-DNN bug. See https://github.com/intel/mkl-dnn/issues/431

----

Update: an updated test program to demo the problem is at https://github.com/pytorch/pytorch/issues/18218#issuecomment-475453948

--

Hi, I have a simple python program as below:

```
  1 import torch
  2
  3 #torch.set_flush_denormal(False)
  4 torch.set_printoptions(profile="full")
  5 torch.set_printoptions(precision=65)
  6
  7 conv2 = torch.load("conv2.pth")
  8 print("haha conv2")
  9 for param in conv2.parameters():
 10     print(param.size())
 11     #print(param.data)
 12
 13 input = torch.load("input.pth")
 14 print("haha input")
 15 print(input[0].size())
 16 #print(input[0])
 17
 18 output = conv2(input[0])
 19 print("haha output")
 20 print(output.size())
 21 print(output)
```

I ran the same code (CPU only, no GPU) on different machines:
machine `A` with `Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz`
machine `B` with `Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz`
machine `C` with `Intel(R) Xeon(R) Platinum 8124M CPU @ 3.00GHz`

While I was loading the same `input.pth` and the same `conv2.pth`, the output from machine `C` is difference with the other machines, the output from machine `A` are the same as the output from machine `B`. I tried the same docker image on different machines, so probably it's not a software problem.

I don't have permission to call `PTRACE_TRACEME` on machine `C`, so ltrace wouldn't help in this case.

Any one could tell me how do I know which low level C library function does pytorch call when computing conv2d?

I tried `objdump -T` on `libmkldnn.so.0` and set breakpoint on `mkldnn_dilated_convolution_forward_desc_init`  and `mkldnn_convolution_forward_desc_init` but it didn't stop, so probably I was guessing the wrong function.

Any hints are greatly appreciated!

**Comment by fracting** [comment]
[conv2_input.zip](https://github.com/pytorch/pytorch/files/2987978/conv2_input.zip)

Attached are the data files to reproduce this problem.

```
$ md5 conv2.pth
MD5 (conv2.pth) = 7ba2908fb86852affa164d77e97a037f
$ md5 input.pth
MD5 (input.pth) = 59b814700fa1dcbbca1f56bce568ba37
```

Last few lines of `good` output are:
```
1646785            -3.38981270790100097656250000000000000000000000000000000000000000000e+00,
1646786            -3.34102034568786621093750000000000000000000000000000000000000000000e+00,
1646787            -1.77292987704277038574218750000000000000000000000000000000000000000e-01]]]],
1646788        grad_fn=<MkldnnConvolutionBackward>)
```

Last few lines of `bad` output are:
```
1646785            -3.38981294631958007812500000000000000000000000000000000000000000000e+00,
1646786            -3.34102010726928710937500000000000000000000000000000000000000000000e+00,
1646787            -1.77292868494987487792968750000000000000000000000000000000000000000e-01]]]],
1646788        grad_fn=<MkldnnConvolutionBackward>)
```

Note the first line in good snip ends with `56250000*` while the first line in bad snip ends with `81250000*`

**Comment by SsnL** [comment]
Different CPUs may use different conv algorithms. Also palatalized floating point arithmetic is not deterministic.

**Comment by soumith** [comment]
as @ssnl said, they are the same to a particular machine precision. Beyond that, different CPU models, thread count, and other factors might give you different results (but same upto machine precision)

**Comment by fracting** [comment]
Thanks @SsnL and @soumith , that was a very important lesson to learn!

Could I ask a follow up question?

We have a deep learning classification model (PyTorch) which predicts the same results for same validation set on 3 different machines, however it's prediction on the `Intel(R) Xeon(R) Platinum 8124M CPU @ 3.00GHz` machine significantly bias to one category. For example, when all other machines outputs `bedroom` category, the Xeon machine outputs `shop`, when all other machines outputs `backyard` category, the Xeon machine still outputs `shop`.

Do you have any suggestions how to track this kind of issue?

Is there any well known floating number test suites which I can run on different machines and compare the result? Or is the PyTorch test suite useful for this kind of diagnosing?

Thank you again!

**Comment by SsnL** [comment]
@fracting It could be a problem with mkldnn. Could you try building from source w/o mkldnn and see if the xeon machine outputs a different prediction?

**Comment by fracting** [comment]
Hi @SsnL 

Thanks for the hints! I compile PyTorch from scratch, either with or without mkldnn (by exporting `USE_MKLDNN="FALSE"`)

I can confirm the model outputs bad prediction with mkldnn, but outputs good prediction without mkldnn. (Both on Xeon)

**Comment by SsnL** [comment]
@soumith This is probably worth taking a look at.  

@fracting Could you update the issue description to include your new findings? A more detailed approach to reproduce the problem using the network & weights would be great. Thanks!

**Comment by soumith** [comment]
this is high-priority, if there's any way to reproduce your issue, we will look and fix mkl-dnn.

cc: @mingfeima

**Comment by Jianhui-Li** [comment]
cc @jgong5

**Comment by mingfeima** [comment]
@soumith we will double check this immediately and reply to you soon.

**Comment by fracting** [comment]
[demo2.tar.gz](https://github.com/pytorch/pytorch/files/2994453/demo2.tar.gz)



Hi, I can reduce the test to the attached code and data.


The code loads a conv2 layer from `conv2.pth`, loads a demo input from `input.pth`, then computes the current output, comparing with `output.i7.pth` and `output.xeon.pth`

According to my test, the cosine similarity between i7 output and xeon output is pretty low:

```
tensor([[[ 2.350216172635555e-02,  1.297117769718170e-01,
           1.381154656410217e-01,  6.210334133356810e-03,
           3.263642266392708e-02,  1.101085636764765e-02,
           2.132313139736652e-02,  5.432983487844467e-02,
```

Env:

Python3.6, with pytorch installed from latest conda


Please let me know if you can reproduce. I'm glad to reduce it to C code if you can assist. Thanks!

**Comment by mingfeima** [comment]
@fracting i see your post on [mkl-dnn](https://github.com/intel/mkl-dnn/issues/431) github, the info for me should be good enough to kick off debugging.
I will let you know if i need additional info.

**Comment by Jianhui-Li** [comment]
@fracting  Between output.i7.pth and output.xeon.pth, both of them are processed on MKL-DNN?

**Comment by fracting** [comment]
Hi, @Jianhui-Li , yes, both output.i7.pth and output.xeon.pth are computed with official conda pytorch package, which has MKL-DNN by default.

Also torch._C.has_mkl are True on both environment.

Hi @mingfeima thanks for working on it! Let me know what ever info is needed.

**Comment by fracting** [comment]
Retested on a Xeon machine with both pytorch-1.0.1 and pytorch-1.1.0 from official conda binary packages, confirming fixed in pytorch-1.1.0.

```
$ cat /proc/cpuinfo | grep flags
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc aperfmperf pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single kaiser fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f rdseed adx smap clflushopt clwb avx512cd xsaveopt xsavec xgetbv1 ida arat pku
```

Closing the issue, thank you for all the great work!
